### DISCUSSION TITLE
Control non-PWM fans?

### DISCUSSION THREAD

**Discussion by chrismyers81** [tjko/fanpico#68] [Q&A] [answered]
Is it possible to control non-PWM fan speeds as well? Thus far I haven't found a config that works for that.

**Comment by tjko** [comment, marked as answer]
Current boards only support controlling speed of PWM (4-pin) fans. Otherwise 3-pin non-PWM fans work (just no speed control).

Controlling speed of non-PWM (fixed speed) fan is possible by adjusting the fan input voltage, but that doesn't seem very efficient.
And it would make the PCB much more complicated and costly, so not sure if it would be worth the effort, as nowadays PWM controlled fans are not much more expensive.

However, if someone is interested in creating new fanpico board design with support for non-PWM fans that would be most welcome.

**Reply by tjko** [reply]
Does anyone happen to know how well it would work to control fixed speed (3-pin) fan speed using PWM (modulated) 12V power? As doing that would be much simpler to implement and require  only few extra components (per fan channel).

**Reply by chrismyers81** [reply]
I haven't tried it personally (was giving it a lot of thought, then decided to just buy new PWM fans.)

But, I will say that several of my computer motherboards do control fan speed with just three-pin headers. I can't say for sure how they work, but my assumption is that it's just PWM on the 12v power to the fan itself.

**Reply by tjko** [reply]
Quick search didn't seem to produce much results if using "PWM modulated" 12V power would work, or would need to have true voltage control...
I need to test with couple fans when I have time. Since if PWM works then it might be feasible to add jumpers on (some) fan ports to allow switching between PWM signal or voltage for speed control...

**Reply by kiler129** [reply]
@tjko I think I can answer some of your questions. 

### Overview
There are a few gotchas:
1. Some fans hate PWM *(rare nowadays)* and will have very narrow range of speed or not operate properly *(very rare)*
2. Many fans have ~~cheap~~ cost-optimized controller, so make sure to use high-frequency PWM
3. Tach signal is tricky due to low-side control


### PWM control
Standard 12V PC fans can be controller with low-side PWM. While DC fan controls exist, it's rare to see pure voltage control for PC fans:
1. Overall implementation cost
2. Complexity of switch-mode regulation
3. Linear regulators being impractical due power requirements vs heat

Using PWM works very well if you ensure PWM frequency of >20kHz. Ideally per spec it's 22.5kHz, but from experience anything in a range of ~20-30kHz works well. 5V 3-pin PC fans largely don't support PWM control, or have almost useless usable control range.


### Tach signal
Since you're creating a floating ground reference, the tach signal will be garbled in-between cycles. Usually how it's solved by controllers is using is a periodic pulse of 100% PWM for a time of ~200% of last rotational speed of the fan. Since tach signal should be 2 pulse/rev, for e.g. 1200rpm it will give you 40 pulses/s or one pulse every ~25ms. In such case, you should set the 100% PWM for ~80-100ms and read the tach signal during that period.
This technique is called "pulse stretching" and AnalogDevices has a nice article on that (Fig 6) and other topics: https://www.analog.com/en/resources/analog-dialogue/articles/how-to-control-fan-speed.html


### Detection of fans
As for 4-pin vs 3-pin detection, this is actually very simple to solve without a jumper. Per specification PWM pin of a 4-pin fan is always pulled up by the fan (i.e. you should have it floating in the controller). This is officially codified in spec published by Intel *"4-Wire Pulse Width Modulation (PWM) Controlled Fans"*. You can get rev 1.3 from 2005, which is I believe the newest one as [PDF from Intel](https://www.intel.com/content/dam/support/us/en/documents/intel-nuc/intel-4wire-pwm-fans-specs.pdf), or [better quality form 3rd party](https://www.glkinst.com/cables/cable_pics/4_Wire_PWM_Spec.pdf).

Since life isn't ideal, some fans have very weak pull-ups and can misbehave esp. when on a longer cable. This is why commercial fan controllers (e.g. ones built-in into motherboards) usually allow users to pick between control schema: Autodetected / DC / 4-pin PWM. That pull-up is usually to ~3.3V, but some fans pull it to ~5V as well.


### More info
There are some older resouces like http://pcbheaven.com/circuitpages/PWM_3_Wires_Fan_Controller_with_RPM_feedback/. You can also see a great [whitepaper from Noctua](https://noctua.at/pub/media/wysiwyg/Noctua_PWM_specifications_white_paper.pdf) on the topic. From experience, Noctua engineers were (are? ;)) quite happy to answer implementation questions.

**Reply by tjko** [reply]
Thanks for the detailed response.  That tachometer signal issue caused by PWM modulated power makes things indeed bit more complicated.  I wonder if there is some good "off the shelf" fan controller/driver ICs that would make sense to use if want to support both 4-pin and 3-pin fans...

I'm not sure if there really would be "demand" to support 3-pin fan control (in FanPico), with the minimal price delta between 3-pin and 4-pin fans?
But if someone wants to create new board with the support, I would gladly help adding support to the new board variant in in the FanPico firmware.
